src/libs/ApiUtils.js

```diff
@@ -0,0 +1,86 @@
+import lodashGet from 'lodash/get';
+import Onyx from 'react-native-onyx';
+import ONYXKEYS from '../ONYXKEYS';
+import CONFIG from '../CONFIG';
+import CONST from '../CONST';
+import * as Environment from './Environment/Environment';
+
+// To avoid rebuilding native apps, native apps use production config for both staging and prod
+// We use the async environment check because it works on all platforms
+let ENV_NAME = CONST.ENVIRONMENT.PRODUCTION;
+let stagingServerToggleState = false;
+Environment.getEnvironment()
+    .then((envName) => {
+        ENV_NAME = envName;
+
+        // We connect here, so we have the updated ENV_NAME when Onyx callback runs
+        Onyx.connect({
+            key: ONYXKEYS.USER,
+            callback: (val) => {
+                const defaultToggleState = ENV_NAME === CONST.ENVIRONMENT.STAGING;
+                stagingServerToggleState = lodashGet(val, 'shouldUseStagingServer', defaultToggleState);
+            },
+        });
+    });
+
+/**
+ * Helper method used to decide which API endpoint to call
+ * Non PROD environments allow API switching via the {@link stagingServerToggleState}
+ *
+ * @returns {Boolean}
+ */
+function canUseStagingServer() {
+    return ENV_NAME === CONST.ENVIRONMENT.PRODUCTION
+        ? false
+        : stagingServerToggleState;
+}
+
+/**
+ * Get the currently used API endpoint
+ * (Non-production environments allow for dynamically switching the API)
+ *
+ * @param {Object} [request]
+ * @param {Boolean} [request.shouldUseSecure]
+ * @returns {String}
+ */
+function getApiRoot(request) {
+    const shouldUseSecure = lodashGet(request, 'shouldUseSecure', false);
+
+    if (canUseStagingServer()) {
+        return shouldUseSecure
+            ? CONFIG.EXPENSIFY.STAGING_SECURE_API_ROOT
+            : CONFIG.EXPENSIFY.STAGING_API_ROOT;
+    }
+
+    return shouldUseSecure
+        ? CONFIG.EXPENSIFY.DEFAULT_SECURE_API_ROOT
+        : CONFIG.EXPENSIFY.DEFAULT_API_ROOT;
+}
+
+/**
+ * Get the command url for the given request
+ *
+ * @param {Object} request
+ * @param {String} request.command - the name of the API command
+ * @param {Boolean} [request.shouldUseSecure]
+ * @returns {String}
+ */
+function getCommandURL(request) {
+    return `${getApiRoot(request)}api?command=${request.command}`;
+}
+
+/**
+ * Check if we're currently using the staging API root
+ *
+ * @returns {Boolean}
+ */
+function isUsingStagingApi() {
```

Comment:
Isn't `isUsingStagingApi` the same as `canUseStagingServer`? It seems redundant.

Reply:
Yeah, this was mainly created for external usages like the `TestToolMenu.js` where an export like `isUsing` - it's a better name compared to `shouldUse` or `canUse` (those are more internal as they disclose how stuff works)

Instead of 

```js
return getApiRoot() === CONFIG.EXPENSIFY.STAGING_API_ROOT;
```

we can 

```js
return canUseStagingServer();
```

But the function would just look weird

```js
function isUsingStagingApi() {
  return canUseStagingServer();
}
```

I don't want to explain why `isUsingStagingApi` `==` `canUseStagingServer`, because `is` is not always `==` `can` 
So I prefer to put a check that is inline with what the function does (and is still a one liner)

If you it's a problem I can remove the function altogether it's all about `is` vs `can` stuff and how it is exported for external usage

Reply:
That's clear, thanks @kidroca but I think the name `canUseStagingServer` is misleading, it should be reset to `shouldUseStagingServer` as that's what the function really do.

Reading `canUseStagingServer` I expected a simple `return ENV_NAME !== CONST.ENVIRONMENT.PRODUCTION` which is not the case.

Reply:
I can rename `canUseStagingServer` to `isUsingStagingApi` and export it, so we can delete the function here, is that alright? 

The logic in `getRootApi` becomes 
- if: `isUsingStagingApi` -> use staging api
- otherwise: use default api

Reply:
I think it would be better to do it this was:
1. Remove the function `canUseStagingServer` given it's unlikely to be used anywhere besides `getApiRoot`
2. On `getApiRoot` define `const shouldUseStagingServer = ENV_NAME === CONST.ENVIRONMENT.PRODUCTION ? false : stagingServerToggleState;`

___

Question: Why `isUsingStagingApi` only checks if the env is staging and not secure-staging?

Reply:
Regarding 1 and 2 - I like to keep conditions simple and easy to understand like `if (shouldUseStaging()` vs the expression you suggest 
Though I can rename `let stagingServerToggleState` and add all the logic in the Onyx callback so we just use value

> Question: Why `isUsingStagingApi` only checks if the env is staging and not secure-staging?

Because it's not possible to use prod api but secure staging api - you always use the APIs together - staging with secure staging

Reply:
But if you are using secure-staging the `getApiRoot` will return `CONFIG.EXPENSIFY.STAGING_SECURE_API_ROOT` and `isUsingStagingApi()` will return false.

Reply:
> But if you are using secure-staging the `getApiRoot` will return `CONFIG.EXPENSIFY.STAGING_SECURE_API_ROOT` and `isUsingStagingApi()` will return false.

Sorry, this doesn't work this way
Whether is a secure request or not you are either using the staging api or the default api 
Depending on the request you'll get either the secure path or the non secure path - but in both cases when the toggle is toggled to staging, it's a staging api url

Reply:
I think you didn't understand my question, let me rephrase, If my env is staging and I'm using the secure path, I'd be using the staging api, right?

If the answer to that question is yes, then why `isUsingStagingApi()` says otherwise (returns false)

Reply:
I don't think we have to discuss this anymore, the logic was changed, is the new logic ok?

Reply:
Oh sorry you fixed that on the last commit too, I didn't notice since those comments covered most of the page bottom and I though it's the end of the changes :sweat_smile:. Looking good now. Thanks :+1: